Source/WebCore/loader/IntegrityPolicy.cpp

```diff
@@ -0,0 +1,142 @@
+            auto* token = std::get_if<RFC8941::Token>(&item);
+            if (!token)
+                continue;
+            auto value = token->string();
+
+            if ((!match || value == *match))
+                vector.append(value);
+        }
+    };
+    // 2. Let dictionary be the result of getting a structured field value from headers given headerName and "dictionary".
+    auto parsedHeader = RFC8941::parseDictionaryStructuredFieldValue(response.httpHeaderField(headerName));
+    if (!parsedHeader)
+        return nullptr;
+    // 1. Let integrityPolicy be a new integrity policy struct.
+    std::unique_ptr<IntegrityPolicy> policy = makeUnique<IntegrityPolicy>();
+    bool sourcesExists = false;
+    for (auto& [name, valueAndParameters] : *parsedHeader) {
+        RFC8941::InnerList* innerList = std::get_if<RFC8941::InnerList>(&valueAndParameters.first);
+        if (!innerList)
+            continue;
+        // 4. If dictionary["blocked-destinations"] exists:
+        // 4.1. If its value contains "script", append "script" to integrityPolicy’s blocked destinations.
+        if (name == "blocked-destinations"_s)
+            parseInnerList(*innerList, policy->blockedDestinations, "script"_s);
+        // 3. If dictionary["sources"] ... value contains "inline", append "inline" to integrityPolicy’s sources.
+        else if (name == "sources"_s) {
+            sourcesExists = true;
+            parseInnerList(*innerList, policy->sources, "inline"_s);
+        } else if (name == "endpoints"_s)
+            parseInnerList(*innerList, policy->endpoints, std::nullopt);
+    }
+    // 3. If dictionary["sources"] does not exist, append "inline" to integrityPolicy’s sources.
+    if (!sourcesExists)
+        policy->sources.append("inline"_s);
+    return policy;
+}
+
+// https://w3c.github.io/webappsec-subresource-integrity/#should-request-be-blocked-by-integrity-policy-section
+bool shouldRequestBeBlockedByIntegrityPolicy(ScriptExecutionContext& context, const ResourceLoaderOptions& options, const URL& url)
+{
+    auto integrityPolicy = context.integrityPolicy();
+    auto integrityPolicyReportOnly = context.integrityPolicyReportOnly();
+
+    // 2. Let parsedMetadata be the result of calling parse metadata with request’s integrity metadata.
+    auto parsedMetadata = parseIntegrityMetadata(options.integrity);
+    // 3. If parsedMetadata is not the empty set and request’s mode is either "cors" or "same-origin", return "Allowed".
+    if ((parsedMetadata && parsedMetadata->size() && options.mode != FetchOptions::Mode::NoCors) || url.protocolIsBlob() || url.protocolIsData())
+        return false;
+    // 6. If both policy and reportPolicy are empty integrity policy structs, return "Allowed".
+    if (!integrityPolicy && !integrityPolicyReportOnly)
+        return false;
+    // We don't currently support anything but script
+    if (options.destination != FetchOptionsDestination::Script)
```

Comment:
I think there is currently support for integrity metadata on HTMLLinkElements for CSS style sheets, and support for integrity for CSS images is in the works. Will this check be problematic for either of those? (I realize the PR is about implementing this for scripts, so I am not asking for you to add more support, just checking that this won't break anything).

Reply:
> and support for integrity for CSS images is in the works

I did not know that. Any pointers?

> Will this check be problematic for either of those? 

It won't, as non-script destinations would just pass this check and then fail at their actual integrity verifications (which happens at response time)

Reply:
Thanks!

One thing to watch out for here is that verifying integrity on images would prevent them from being progressively decoded, which may result in performance regressions.